room/room-compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KspBasicAnnotationProcessor.kt

```diff
@@ -44,16 +44,25 @@ abstract class KspBasicAnnotationProcessor(
             symbolProcessorEnvironment.logger
         )
         val round = XRoundEnv.create(processingEnv)
-        val deferredElements = processingSteps().flatMap { step ->
-            val invalidElements = mutableSetOf<XElement>()
-            val elementsByAnnotation = step.annotations().associateWith { annotation ->
+        val deferredElements = mutableSetOf<XElement>()
+        processingSteps().forEach { step ->
+            val elementsByAnnotation = mutableMapOf<String, Set<XElement>>()
+            step.annotations().forEach { annotation ->
                 val annotatedElements = round.getElementsAnnotatedWith(annotation)
-                annotatedElements
-                    .filter { (it as KspElement).declaration.validateExceptLocals() }
-                    .also { invalidElements.addAll(annotatedElements - it) }
-                    .toSet()
+                val validAnnotatedElements =
+                    annotatedElements
+                        .filter { (it as KspElement).declaration.validateExceptLocals() }
+                        .toSet()
+
+                deferredElements.addAll(annotatedElements - validAnnotatedElements)
+                if (!validAnnotatedElements.isEmpty()) {
+                    elementsByAnnotation[annotation] = validAnnotatedElements
+                }
+            }
+            // Only process the step if there are annotated elements found for this step.
```

Comment:
I think we can do this without increasing the scope of a mutable deferred elements, we need to change the `associateWith` to a `mapNotNull` of `Pair` where `null` is no annotated elements and then make that a map, it should look like this:
```
val deferredElements = processingSteps().flatMap { step ->
    val invalidElements = mutableSetOf<XElement>()
    val elementsByAnnotation = step.annotations().mapNotNull { annotation ->
        val annotatedElements = round.getElementsAnnotatedWith(annotation)
        val validElements = annotatedElements
            .filter { (it as KspElement).declaration.validateExceptLocals() }
            .also { invalidElements.addAll(annotatedElements - it) }
            .toSet()
        if (validElements.isNotEmpty()) {
            annotation to validElements
        } else {
            null
        }
    }.toMap()
    // Only process the step if there are annotated elements found for this step.
    if (elementsByAnnotation.isNotEmpty()) {
        invalidElements + step.process(processingEnv, elementsByAnnotation)
    } else {
        invalidElements
    }
}
```

Reply:
Done.